File: pa3/pa3q2.py
import sys

knapitems = []
with open('knapsack_big.txt', 'r') as f:
    knapsize, numitems = [int(x) for x in f.readline().split(' ')]
    minweight = 100000000000
    for line in f:
        curval, curweight = [int(x) for x in line.split(' ')]
        minweight = min(minweight, curweight)
        curitem = (curval, curweight)
        knapitems.append(curitem)
    f.close()

sys.setrecursionlimit(max(knapsize, numitems)*10)
A = {}


def calcKnapsack(itemno, weight):
    if itemno <= -1 or weight < minweight:
        return 0
    curstr = str(itemno) + "," + str(weight)
    if curstr in A.keys():
        return A[curstr]
    else:
        curval = knapitems[itemno][0]
        curweight = knapitems[itemno][1]
        newitemno = itemno-1
        term1str = str(newitemno) + "," + str(weight)
        if term1str in A.keys():
            term1 = A[term1str]
        else:
            term1 = calcKnapsack(newitemno, weight)
            A[term1str] = term1
        if curweight > weight:
            A[curstr] = term1
            return term1
        else:
            newweight = weight-curweight
            term2str = str(newitemno) + "," + str(newweight)
            if term2str in A.keys():
                term2 = A[term2str]
            else:
                term2 = calcKnapsack(newitemno, newweight) + curval
                A[term2str] = term2
            retval = max(term1, term2)
            A[curstr] = retval
            return retval

returnval = calcKnapsack(numitems-1, knapsize)
print(returnval)
        
        

File: pa3/pa3q2.lua
--lua5.2
local res = {}
local optimal
local knapitems = {}
local calcknap
local knapsize
local knapwt
calcknap = function (itemno, curwt)
    if itemno < 1 then
        return 0
    end
    local curstr = tostring(itemno)..","..tostring(curwt)
    if res[curstr] ~= nil then
        return res[curstr]
    else
        local itemval = knapitems[itemno]["val"]
        local itemwt = knapitems[itemno]["wt"]
        local newitemno = itemno - 1
        local str1 = tostring(newitemno)..","..tostring(curwt)
        local term1
        if res[str1] ~= nil then
            term1 = res[str1]
        else
            term1 = calcknap(newitemno, curwt)
            res[str1] = term1
        end
        if curwt < itemwt then
            return term1
        else
            local term2
            local newwt = curwt - itemwt
            local str2 = tostring(newitemno)..","..tostring(newwt)
            if res[str2] ~= nil then
                term2 = res[str2]
            else
                term2 = calcknap(newitemno, newwt)
                res[str2] = term2
            end
            local retval = math.max(term1,term2+itemval)
            res[curstr] = retval
            return retval
        end
    end
end

local split = "[^%s]+"
local file = "knapsack_big.txt"
local lineno = 0
for line in io.lines(file) do
    local curnum = 1
    local curitem = {}
    for val in string.gmatch(line, split) do
        curitem[curnum] = val
        curnum = curnum + 1
    end
    if lineno == 0 then
        knapwt = tonumber(curitem[1])
        knapsize = tonumber(curitem[2])
    else
        local itemval = tonumber(curitem[1])
        local itemwt = tonumber(curitem[2])
        local toadd = {val=itemval,wt=itemwt}
        knapitems[lineno] = toadd
    end
    --print(lineno)
    lineno = lineno + 1
end

optimal = calcknap(knapsize, knapwt)
print(optimal)
